{"cells":[{"cell_type":"markdown","metadata":{"id":"0atUDr9OfcGi"},"source":["# Excepciones en Python 3"],"id":"0atUDr9OfcGi"},{"cell_type":"markdown","metadata":{"id":"Wiyk954ufcGi"},"source":["### 1. ¿Qué son las excepciones?"],"id":"Wiyk954ufcGi"},{"cell_type":"markdown","metadata":{"id":"AKCavRyHfcGj"},"source":["Python utiliza un tipo de objeto especial denominado excepción para gestionar los errores que surgen durante la ejecución de un programa.\n","\n","Cada vez que ocurre un error que hace que Python no sepa qué hacer a continuación, crea un objeto de excepción. Si se escribe código que maneja la excepción, el programa continuará ejecutándose. Por el contrario, si no se maneja la excepción, el programa se detendrá y mostrará un pequeño resumen de la excepción que se ha producido.\n","\n","Las excepciones se pueden manejan a través de las sentencias `try` y `except`."],"id":"AKCavRyHfcGj"},{"cell_type":"code","execution_count":1,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"LKY35xugfcGk","executionInfo":{"status":"ok","timestamp":1698332311530,"user_tz":-120,"elapsed":12,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"49bedb30-fcad-4b8e-9092-03fff38ba769"},"outputs":[{"output_type":"stream","name":"stdout","text":["variable no declarada\n"]}],"source":["flag = True\n","try:\n","  print(carlos)\n","except:\n","  print(\"variable no declarada\")"],"id":"LKY35xugfcGk"},{"cell_type":"markdown","metadata":{"id":"I84ppLsjfcGk"},"source":["### 2. Errores de sintaxis vs Excepciones"],"id":"I84ppLsjfcGk"},{"cell_type":"markdown","metadata":{"id":"mR2v59h0fcGl"},"source":["A pesar de que en muchas ocasiones estos dos términos se utilizan de manera indistinta cuando hablamos de Python, debemos tener cuidado porque son cosas diferentes.\n","\n","Los **errores de sintaxis** se producen cuando escribimos una sentencia de código en Python que no es sintácticamente válida. El intérprete de Python indica estos error con el término `SyntaxError` y nos señala con el carácter `^` donde se encuentra el error."],"id":"mR2v59h0fcGl"},{"cell_type":"code","execution_count":5,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":142},"id":"-BIQodLafcGl","executionInfo":{"status":"error","timestamp":1698332380489,"user_tz":-120,"elapsed":269,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"79c726a8-eccf-4d57-c4a8-586e246285e3"},"outputs":[{"output_type":"error","ename":"SyntaxError","evalue":"ignored","traceback":["\u001b[0;36m  File \u001b[0;32m\"<ipython-input-5-62a1fbc8656d>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    print(\"Hola mundo)\u001b[0m\n\u001b[0m          ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m unterminated string literal (detected at line 1)\n"]}],"source":["print(\"Hola mundo)"],"id":"-BIQodLafcGl"},{"cell_type":"markdown","metadata":{"id":"vKXqmNBmfcGm"},"source":["**Este tipo de errores no se pueden controlar y no se corresponden con excepciones dentro del lenguaje**"],"id":"vKXqmNBmfcGm"},{"cell_type":"markdown","metadata":{"id":"emtFTnDifcGn"},"source":["Por otro lado, las **excepciones** se corresponden con errores que se producen en sentencias de código en Python que son sintácticamente correctas. Esto tipo de errores no son fatales para el programa y pueden ser gestionados o ignorados."],"id":"emtFTnDifcGn"},{"cell_type":"code","execution_count":6,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":176},"id":"Ywa3t4RRfcGo","executionInfo":{"status":"error","timestamp":1698332416440,"user_tz":-120,"elapsed":10,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"27399a94-927e-42c9-c7d4-e0b58a398add"},"outputs":[{"output_type":"error","ename":"NameError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m<ipython-input-6-2ea387ab95ff>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvar\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mNameError\u001b[0m: name 'var' is not defined"]}],"source":["print(var)"],"id":"Ywa3t4RRfcGo"},{"cell_type":"code","execution_count":8,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":176},"id":"UTgDmbhUfcGo","executionInfo":{"status":"error","timestamp":1698332419625,"user_tz":-120,"elapsed":19,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"8f62ff72-a4dc-4f93-ebad-b5ffd4aa39cf"},"outputs":[{"output_type":"error","ename":"ZeroDivisionError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-8-ff5d2f231fa4>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;36m50\u001b[0m\u001b[0;34m/\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mZeroDivisionError\u001b[0m: division by zero"]}],"source":["50/0"],"id":"UTgDmbhUfcGo"},{"cell_type":"code","execution_count":7,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":176},"id":"Xlm17gHhfcGo","executionInfo":{"status":"error","timestamp":1698332418060,"user_tz":-120,"elapsed":244,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"d5383b8a-3f13-483e-aced-f3a3a96fadbd"},"outputs":[{"output_type":"error","ename":"TypeError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m<ipython-input-7-d2b23a1db757>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;34m'2'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mTypeError\u001b[0m: can only concatenate str (not \"int\") to str"]}],"source":["'2' + 2"],"id":"Xlm17gHhfcGo"},{"cell_type":"markdown","metadata":{"id":"vg6FdzjTfcGo"},"source":["### 3. Gestión de excepciones: Sentencias `try` y `except`"],"id":"vg6FdzjTfcGo"},{"cell_type":"markdown","metadata":{"id":"Iv8Zek0ofcGp"},"source":["Las sentencias `try` y `except` en Python pueden utilizarse para capturar y manejar excepciones. Python ejecuta el código que sigue a la sentencia `try` como una parte \"normal\" del programa. El código que sigue a la sentencia `except` se ejecutará si se produce cualquier excepción en la cláusula `try` precedente.\n","\n","La sintaxis que se utiliza para definir este tipo de comportamiento es la siguiente:\n","```\n","try:\n","    <sentencia(s)>\n","except <excepción>:\n","    <sentencias(s) si excepción>\n","```"],"id":"Iv8Zek0ofcGp"},{"cell_type":"code","execution_count":13,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"cUCuAHagfcGp","executionInfo":{"status":"ok","timestamp":1698332601693,"user_tz":-120,"elapsed":437,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"8a2a3c4d-b17c-43b0-c80a-f4aca7efaf22"},"outputs":[{"output_type":"stream","name":"stdout","text":["qué es var?\n"]}],"source":["try:\n","  print(var)\n","except NameError:\n","  print(\"qué es var?\")"],"id":"cUCuAHagfcGp"},{"cell_type":"code","execution_count":10,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"0N3QmoVpfcGq","executionInfo":{"status":"ok","timestamp":1698332462733,"user_tz":-120,"elapsed":288,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"f0d99287-e6bb-4815-f126-ff97b329ea57"},"outputs":[{"output_type":"stream","name":"stdout","text":["pecado para las mates\n"]}],"source":[],"id":"0N3QmoVpfcGq"},{"cell_type":"code","execution_count":11,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"nia0DUJVfcGq","executionInfo":{"status":"ok","timestamp":1698332489348,"user_tz":-120,"elapsed":230,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"c7e69db3-a805-4c76-9b61-19e6da1f63ec"},"outputs":[{"output_type":"stream","name":"stdout","text":["costaba más poner las comillas que dejarselas\n"]}],"source":["try:\n","  2+'2'\n","except TypeError:\n","  print(\"costaba más poner las comillas que dejarselas\")"],"id":"nia0DUJVfcGq"},{"cell_type":"markdown","metadata":{"id":"RStDdvQ3fcGr"},"source":["Debemos tener en cuenta que el las sentencias de código que se encuentren en el cuerpo de la sentencia `try` y a continuación de la sentencia que emite la excepción, no se ejecutarán"],"id":"RStDdvQ3fcGr"},{"cell_type":"code","execution_count":14,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"cwKFRZXIfcGr","executionInfo":{"status":"ok","timestamp":1698332628354,"user_tz":-120,"elapsed":10,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"359da24e-2e5a-48ec-cf14-421234e5f429"},"outputs":[{"output_type":"stream","name":"stdout","text":["carlos\n","hola carlos\n"]}],"source":["try:\n","  print(\"carlos\")\n","  print(carlos)\n","  print(\"too much carlos\")\n","except NameError:\n","  print(\"hola carlos\")"],"id":"cwKFRZXIfcGr"},{"cell_type":"markdown","metadata":{"id":"NlgkQL0SfcGr"},"source":["También podemos utilizar la sentencia `except` sin indicarle el nombre de ninguna excepción, en estos casos capturará todas las excepciones que se produzcan en el código que se encuentra en el cuerpo de la sentencia `try`"],"id":"NlgkQL0SfcGr"},{"cell_type":"code","execution_count":15,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"okmq7UuufcGr","executionInfo":{"status":"ok","timestamp":1698332629818,"user_tz":-120,"elapsed":235,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"7439fe63-6015-4f35-e04d-81a0cabd3e7b"},"outputs":[{"output_type":"stream","name":"stdout","text":["pecado para las mates\n"]}],"source":["try:\n","  50/0\n","except :\n","  print(\"pecado para las mates\")"],"id":"okmq7UuufcGr"},{"cell_type":"markdown","metadata":{"id":"OhK4XKRKfcGr"},"source":["Adicionalmente a la sintaxis anterior, podemos capturar varias excepciones de manera simultánea utilizando varias cláusulas `except`"],"id":"OhK4XKRKfcGr"},{"cell_type":"code","execution_count":16,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"EL49q0UXfcGs","executionInfo":{"status":"ok","timestamp":1698332674300,"user_tz":-120,"elapsed":229,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"23b0df98-362c-420c-e53f-9933c6f1a6a5"},"outputs":[{"output_type":"stream","name":"stdout","text":["que es carlos\n"]}],"source":["try:\n","  print(carlos)\n","  print(2+'2')\n","except NameError:\n","  print(\"que es carlos\")\n","except TypeError:\n","  print(\"no aprendes\")"],"id":"EL49q0UXfcGs"},{"cell_type":"markdown","metadata":{"id":"-IdypCprfcGs"},"source":["Por último, podemos asignar el objeto excepción capturado a una variable y utilizarlos para mostrar más información al respecto."],"id":"-IdypCprfcGs"},{"cell_type":"code","execution_count":20,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"BFpQhiuGfcGs","executionInfo":{"status":"ok","timestamp":1698332779807,"user_tz":-120,"elapsed":13,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"a3e3bd09-7dc4-41af-fd0a-c6f71f004ec3"},"outputs":[{"output_type":"stream","name":"stdout","text":["este es tu erorr:  <class 'NameError'>\n"]}],"source":["try:\n","  print(carlos)\n","  print(2+'2')\n","except NameError as error:\n","  print(\"este es tu erorr: \", type(error))"],"id":"BFpQhiuGfcGs"},{"cell_type":"markdown","metadata":{"id":"xixv_148fcGs"},"source":["### 4. Lanzando excepciones personalizadas"],"id":"xixv_148fcGs"},{"cell_type":"markdown","metadata":{"id":"nQ7e_OwIfcGt"},"source":["Además de las sentencias anteriores que podemos utilizar para controlar excepciones, Python nos proporciona la sentencia `raise` con la que podemos emitir nuestras propias excepciones. Para ello, debemos utilizar la clase por defecto de Python `Exception`"],"id":"nQ7e_OwIfcGt"},{"cell_type":"code","execution_count":21,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"WXm-oLCsfcGt","executionInfo":{"status":"ok","timestamp":1698332797674,"user_tz":-120,"elapsed":292,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"a1f1824a-fb96-49f0-91a2-f6eae35a48d3"},"outputs":[{"output_type":"stream","name":"stdout","text":["Help on class Exception in module builtins:\n","\n","class Exception(BaseException)\n"," |  Common base class for all non-exit exceptions.\n"," |  \n"," |  Method resolution order:\n"," |      Exception\n"," |      BaseException\n"," |      object\n"," |  \n"," |  Built-in subclasses:\n"," |      ArithmeticError\n"," |      AssertionError\n"," |      AttributeError\n"," |      BufferError\n"," |      ... and 15 other subclasses\n"," |  \n"," |  Methods defined here:\n"," |  \n"," |  __init__(self, /, *args, **kwargs)\n"," |      Initialize self.  See help(type(self)) for accurate signature.\n"," |  \n"," |  ----------------------------------------------------------------------\n"," |  Static methods defined here:\n"," |  \n"," |  __new__(*args, **kwargs) from builtins.type\n"," |      Create and return a new object.  See help(type) for accurate signature.\n"," |  \n"," |  ----------------------------------------------------------------------\n"," |  Methods inherited from BaseException:\n"," |  \n"," |  __delattr__(self, name, /)\n"," |      Implement delattr(self, name).\n"," |  \n"," |  __getattribute__(self, name, /)\n"," |      Return getattr(self, name).\n"," |  \n"," |  __reduce__(...)\n"," |      Helper for pickle.\n"," |  \n"," |  __repr__(self, /)\n"," |      Return repr(self).\n"," |  \n"," |  __setattr__(self, name, value, /)\n"," |      Implement setattr(self, name, value).\n"," |  \n"," |  __setstate__(...)\n"," |  \n"," |  __str__(self, /)\n"," |      Return str(self).\n"," |  \n"," |  with_traceback(...)\n"," |      Exception.with_traceback(tb) --\n"," |      set self.__traceback__ to tb and return self.\n"," |  \n"," |  ----------------------------------------------------------------------\n"," |  Data descriptors inherited from BaseException:\n"," |  \n"," |  __cause__\n"," |      exception cause\n"," |  \n"," |  __context__\n"," |      exception context\n"," |  \n"," |  __dict__\n"," |  \n"," |  __suppress_context__\n"," |  \n"," |  __traceback__\n"," |  \n"," |  args\n","\n"]}],"source":["help(Exception)"],"id":"WXm-oLCsfcGt"},{"cell_type":"code","execution_count":22,"metadata":{"id":"crFWojyNfcGt","executionInfo":{"status":"ok","timestamp":1698332810564,"user_tz":-120,"elapsed":8,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}}},"outputs":[],"source":["colores_permitidos = (\"azul\", \"verde\", \"rojo\")"],"id":"crFWojyNfcGt"},{"cell_type":"code","execution_count":24,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":211},"id":"CHASG_B2fcGt","executionInfo":{"status":"error","timestamp":1698332844688,"user_tz":-120,"elapsed":240,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"8aa1bc52-80c9-4fb7-da68-1ed5eb54d5c1"},"outputs":[{"output_type":"error","ename":"Exception","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mException\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m<ipython-input-24-6d3efc0721b9>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mcolor\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"amarillo\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mcolor\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcolores_permitidos\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m   \u001b[0;32mraise\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"El color {} no se encuentra en la lista de colores permitidos\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"amarillo\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mException\u001b[0m: El color amarillo no se encuentra en la lista de colores permitidos"]}],"source":["color = \"amarillo\"\n","if color not in colores_permitidos:\n","  raise Exception(\"El color {} no se encuentra en la lista de colores permitidos\".format(\"amarillo\"))"],"id":"CHASG_B2fcGt"},{"cell_type":"markdown","metadata":{"id":"sX67ER_OfcGt"},"source":["### 5. Excepción `AssertionError`"],"id":"sX67ER_OfcGt"},{"cell_type":"markdown","metadata":{"id":"z64nrN5JfcGt"},"source":["Como complemento a todas las sentencias anteriores, Python nos proporciona una sentencia adicional que nos permite verificar en un punto determinado de nuestro programa que todo esta funcionando adecuadamente, esta sentencia es `assert`."],"id":"z64nrN5JfcGt"},{"cell_type":"code","execution_count":26,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":211},"id":"32oiYURffcGu","executionInfo":{"status":"error","timestamp":1698332879799,"user_tz":-120,"elapsed":1980,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"e3e5dcef-226c-4d36-a7cd-2c56c7de5cf3"},"outputs":[{"name":"stdout","output_type":"stream","text":["Introduce una contrasena de mas de 8 digitos:2\n"]},{"output_type":"error","ename":"AssertionError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)","\u001b[0;32m<ipython-input-26-79ef2bab0e22>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mpasswd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Introduce una contrasena de mas de 8 digitos:\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32massert\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpasswd\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m8\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"La contrasena es menor a 8 digitos\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mAssertionError\u001b[0m: La contrasena es menor a 8 digitos"]}],"source":["passwd = input(\"Introduce una contrasena de mas de 8 digitos:\")\n","assert len(passwd) > 8, \"La contrasena es menor a 8 digitos\""],"id":"32oiYURffcGu"},{"cell_type":"markdown","metadata":{"id":"3bQnK5y6fcGu"},"source":["### 6. Cláusula `else` en excepciones"],"id":"3bQnK5y6fcGu"},{"cell_type":"markdown","metadata":{"id":"yrHtsqUVfcGu"},"source":["Curiosamente, Python nos proporciona un mecanismo por el cual utilizando la sentencia `else`, se puede indicar a un programa que ejecute un determinado bloque de código sólo en ausencia de excepciones."],"id":"yrHtsqUVfcGu"},{"cell_type":"code","execution_count":28,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"fAixijRxfcGu","executionInfo":{"status":"ok","timestamp":1698332909572,"user_tz":-120,"elapsed":6,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"f13bc462-6b72-4f68-a900-e0bffc1195b9"},"outputs":[{"output_type":"stream","name":"stdout","text":["variable2 no estaba definida, se define con la cadena 'Hola mundo'\n"]}],"source":["try:\n","  print(variable2)\n","except NameError:\n","  print(\"variable2 no estaba definida, se define con la cadena 'Hola mundo'\")\n","  variable2 = 'Hola mundo'\n","else:\n","  print(\"variable2 ya estaba definida con el valor:\", variable2)"],"id":"fAixijRxfcGu"},{"cell_type":"markdown","metadata":{"id":"pqop4kgufcGu"},"source":["### 7. Sentencia `finally`"],"id":"pqop4kgufcGu"},{"cell_type":"markdown","metadata":{"id":"-mABONuOfcGu"},"source":["Python nos proporciona una última sentencia que podemos utilizar para realizar una \"limpieza\" después de la ejecución de nuestro código al gestionar una excepción. Esta sentencia se denomina `finally` y el código que se localice en su cuerpo, se ejcutará siempre, independientemente de si se produce o no la expcepción."],"id":"-mABONuOfcGu"},{"cell_type":"code","execution_count":29,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Kq9qFl3JfcGv","executionInfo":{"status":"ok","timestamp":1698332943979,"user_tz":-120,"elapsed":248,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"a84420d0-bb0c-42fe-e070-faa8c6246d16"},"outputs":[{"output_type":"stream","name":"stdout","text":["variable3 no estaba definida, se define con la cadena 'Hola mundo'\n"]}],"source":["try:\n","  print(variable3)\n","except NameError:\n","  print(\"variable3 no estaba definida, se define con la cadena 'Hola mundo'\")\n","  variable3 = 'Hola mundo'\n","else:\n","  print(\"variable3 ya estaba definida con el valor:\", variable2)"],"id":"Kq9qFl3JfcGv"},{"cell_type":"code","execution_count":30,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"YFiaFZ7SfcGv","executionInfo":{"status":"ok","timestamp":1698332963953,"user_tz":-120,"elapsed":13,"user":{"displayName":"Carlos Navarro Segarra","userId":"08442718469004035589"}},"outputId":"0d467494-f3a3-42be-d9a6-c10d6c5f8c16"},"outputs":[{"output_type":"stream","name":"stdout","text":["Hola mundo\n","variable3 ya estaba definida con el valor: Hola mundo\n"]}],"source":["try:\n","  print(variable3)\n","except NameError:\n","  print(\"variable3 no estaba definida, se define con la cadena 'Hola mundo'\")\n","  variable3 = 'Hola mundo'\n","else:\n","  print(\"variable3 ya estaba definida con el valor:\", variable2)\n","finally:\n","  del variable3"],"id":"YFiaFZ7SfcGv"},{"cell_type":"code","execution_count":null,"metadata":{"id":"TyTNdIFnfcGv"},"outputs":[],"source":[],"id":"TyTNdIFnfcGv"},{"cell_type":"code","execution_count":null,"metadata":{"id":"hbym_vnifcGv"},"outputs":[],"source":[],"id":"hbym_vnifcGv"}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.5"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}